File: cp/LC1743-Restore-the-Array-From-Adjacent-Pairs.py
from typing import List
from collections import defaultdict


class Solution:
    def restoreArray(self, adjacentPairs: List[List[int]]) -> List[int]:
        graph = defaultdict(list)

        for x, y in adjacentPairs:
            graph[x].append(y)
            graph[y].append(x)

        root = None
        for node in graph:
            if len(graph[node]) == 1:
                root = node
                break

        nums = []

        def dfs(node, prev):
            nums.append(node)

            for nb in graph[node]:
                if nb != prev:
                    dfs(nb, node)

        dfs(root, None)
        return nums


def main():
    n = int(input())
    # adjPairs = [[int(input()) for x in range(2)] for _ in range(n)]
    adjPairs = []
    for _ in range(n):
        row = list(map(int, input().strip().split()))
        adjPairs.append(row)
    sol = Solution()
    nums = sol.restoreArray(adjPairs)

    for num in nums:
        print(num, end=" ")
    print()


if __name__ == "__main__":
    main()

File: cp/LC1743-Restore-the-Array-From-Adjacent-Pairs.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define mod 1000000007

class Solution
{
    void dfs(int node, int prev, unordered_map<int, vector<int>> &graph, vector<int> &nums)
    {
        nums.push_back(node);
        for (auto &nb : graph[node])
        {
            if (nb != prev)
            {
                dfs(nb, node, graph, nums);
            }
        }
    }

public:
    vector<int> restoreArray(vector<vector<int>> &adjacentPairs)
    {
        unordered_map<int, vector<int>> graph;
        for (auto &edge : adjacentPairs)
        {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        int root = 0;
        for (auto &pr : graph)
        {
            if (pr.second.size() == 1)
            {
                root = pr.first;
                break;
            }
        }
        vector<int> nums;
        dfs(root, INT_MAX, graph, nums);
        return nums;
    }
};

int main()
{
    int n;
    cin >> n;
    vector<vector<int>> adjPairs;
    for (int i = 0; i < n; i++)
    {
        int x, y;
        cin >> x >> y;
        adjPairs.push_back({x, y});
    }

    Solution sol;
    vector<int> nums = sol.restoreArray(adjPairs);
    for (auto &num : nums)
    {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}